zerver/lib/digest.py

```diff
@@ -263,3 +264,45 @@ def handle_digest_email(user_profile_id: int, cutoff: float,
         send_future_email('zerver/emails/digest', user_profile.realm, to_user_ids=[user_profile.id],
                           from_name="Zulip Digest", from_address=FromAddress.NOREPLY, context=context)
     return None
+
+def gather_soft_deactivated_user_stream_messages(user_profile: UserProfile,
+                                                 stream_ids: Sequence[int],
+                                                 cutoff_date: datetime.datetime) -> Sequence[Message]:
+    """Gather messages in streams where the soft deactivated user is subscribed.
+
+    Only the following streams are used to gather these messages:
+
+    1. The user should be subscribed to the stream when this code is run.
+
+    2. The user's subscription to the stream shouldn't have changed during the
+    period for which the digest is being generated.
+
+    We query for all the messages sent after the cut-off date, in the streams
+    satisfying the criteria above.
+
+    """
+
+    events = [
+        RealmAuditLog.SUBSCRIPTION_CREATED,
+        RealmAuditLog.SUBSCRIPTION_ACTIVATED,
+        RealmAuditLog.SUBSCRIPTION_DEACTIVATED
+    ]
+
+    # Streams where the user's subscription was changed
+    modified_streams = RealmAuditLog.objects.filter(
+        realm=user_profile.realm,
+        modified_user=user_profile,
+        event_time__gt=cutoff_date,
+        event_type__in=events).values_list('modified_stream_id', flat=True)
+
+    # Only look for streams where the user is currently subscribed, and
+    # subscription didn't change after the cutoff date.
+    streams = set(stream_ids) - set(modified_streams)
+
+    # Fetch list of all messages sent after cutoff_date where the user is subscribed
+    messages = Message.objects.filter(
+        recipient__type=Recipient.STREAM,
+        recipient__type_id__in=streams,
+        pub_date__gt=cutoff_date).select_related('recipient', 'sender', 'sending_client')
+
+    return messages
```

Comment:
So, after reading this, I think that we could have a cleaner setup where this function return a set of `recipient_id` values, and just replaces the `home_view_recipients` argument used above (basically avoiding the `Message.objects.filter()` part of the code here being duplicated with the main code path).

Reply:
I've refactored this. Thanks for the suggestion!